File: src/sell-partition.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub partition {
    my ($n, $templist) = @_;
    my $partlist;

    $templist = { 1 => [[1]] } unless $templist;

    return $templist->{$n} if exists $templist->{$n};

    $partlist = [[$n]];

    foreach (1 .. $n-1) {
        my $rest = $n - $_;

        foreach my $part (@{partition ($_, $templist)}) {
            push @$partlist, [$rest, @$part] if $part->[0] <= $_;
        }
    }

    $templist->{$n} = $partlist;
    return $partlist;
}

sub seller {
    my $prices = shift;

    my @parts = keys %{{
        map { $_ => 1 }
        map { join ' ', @$_ }
        map { [sort @$_] }
        @{partition scalar @$prices}
    }}; # removes duplicate partitions

    my $max = 0;
    foreach (@parts) {
        my @elems = split;

        my $sum = 0;
        $sum += $prices->[$_ - 1] foreach @elems;

        $max = $sum if $sum > $max;
    }

    return $max;
}

my $num_input = <>;

my @input;
foreach (1 .. $num_input) {
    my $input = <>;
    chomp $input;
    push @input, $input;
}

print "${\seller \@input}\n";

File: src/sell-partition.c
#include <stdio.h>
#include <stdlib.h>

int *prices;

int benefit(int *arr, int len)
{
    static int max = 0;
    int sum;

    for (sum = 0; len--; ++arr) sum += prices[*arr - 1];
    if (sum > max) max = sum;

    return max;
}

void partition(int n)
{
    int *part, last, rest;

    part = malloc(n * sizeof(*part));

    part[last = 0] = n;

    for (;;) {
        benefit(part, last + 1);

        rest = 0;

        while (part[last] == 1) {
            rest += part[last--];
            if (last < 0) return;
        }
 
        --part[last];
        ++rest;

        while (rest > part[last]) {
            part[last + 1] = part[last];
            rest -= part[last + 1];
            ++last;
        }

        part[++last] = rest;
    }
}

int seller(int len, int *arr)
{
    prices = arr;
    partition(len);
    return benefit(NULL, 0);
}

int main()
{
    size_t i, n;
    int *arr;

    scanf("%zu\n", &n);
    arr = malloc(n * sizeof(*arr));
    for (i = 0; i < n; ++i) scanf(" %d%*c", arr + i);

    printf("%d\n", seller(8, arr));

    return 0;
}
